http-client-javanet/src/main/java/io/micronaut/http/client/javanet/JavanetHttpClient.java

```diff
@@ -0,0 +1,150 @@
+        @Nullable String contextPath,
+        MediaTypeCodecRegistry mediaTypeCodecRegistry,
+        RequestBinderRegistry requestBinderRegistry,
+        String clientId,
+        ConversionService conversionService,
+        JavanetClientSslBuilder sslBuilder
+    ) {
+        super(LOG, loadBalancer, httpVersion, configuration, contextPath, mediaTypeCodecRegistry, requestBinderRegistry, clientId, conversionService, sslBuilder);
+    }
+
+    public JavanetHttpClient(URI uri, ConversionService conversionService) {
+        this(
+            uri == null ? null : LoadBalancer.fixed(uri),
+            null,
+            new DefaultHttpClientConfiguration(),
+            null,
+            createDefaultMediaTypeRegistry(),
+            new DefaultRequestBinderRegistry(conversionService),
+            null,
+            conversionService,
+            new JavanetClientSslBuilder(new ResourceResolver())
+        );
+    }
+
+    public JavanetHttpClient(
+        URI uri,
+        HttpClientConfiguration configuration,
+        MediaTypeCodecRegistry mediaTypeCodecRegistry,
+        ConversionService conversionService
+    ) {
+        this(
+            uri == null ? null : LoadBalancer.fixed(uri),
+            null,
+            configuration,
+            null,
+            mediaTypeCodecRegistry,
+            new DefaultRequestBinderRegistry(conversionService),
+            null,
+            conversionService,
+            new JavanetClientSslBuilder(new ResourceResolver())
+        );
+    }
+
+    private static MediaTypeCodecRegistry createDefaultMediaTypeRegistry() {
+        JsonMapper mapper = JsonMapper.createDefault();
+        ApplicationConfiguration configuration = new ApplicationConfiguration();
+        return MediaTypeCodecRegistry.of(
+            new JsonMediaTypeCodec(mapper, configuration, null),
+            new JsonStreamMediaTypeCodec(mapper, configuration, null)
+        );
+    }
+
+    @Override
+    public BlockingHttpClient toBlocking() {
+        return new JavanetBlockingHttpClient(loadBalancer, httpVersion, configuration, contextPath, mediaTypeCodecRegistry, requestBinderRegistry, clientId, conversionService, sslBuilder);
+    }
+
+    @Override
+    public <I, O, E> Publisher<HttpResponse<O>> exchange(@NonNull HttpRequest<I> request, @NonNull Argument<O> bodyType, @NonNull Argument<E> errorType) {
+        return mapToHttpRequest(request, bodyType)
+            .map(httpRequest -> {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Client {} Sending HTTP Request: {}", clientId, httpRequest);
+                }
+                if (LOG.isTraceEnabled()) {
+                    httpRequest.headers().map().forEach((k, v) -> LOG.trace("Client {} Sending HTTP Request Header: {}={}", clientId, k, v));
+                }
+                return client.sendAsync(httpRequest, java.net.http.HttpResponse.BodyHandlers.ofByteArray());
+            })
+            .flatMap(Mono::fromFuture)
```

Comment:
this is not idea because `fromFuture` will block waiting for the future to complete. You want to use `fromCompletionStage' instead which can register a listener. Also to support the different `exchangeStream` use cases of the HTTP client you will need to use `HttpResponse.BodyHandlers.ofPublisher()` and include the `Flow.Publisher` adapters.

Reply:
Good knowledge on completionStage vs Future.  Fixed that in https://github.com/micronaut-projects/micronaut-core/pull/8441/commits/72431a66a6a3b6c2d22a9c68235f5f045853b1d4

I'm not sure of the second part?  How do I "_…include the `Flow.Publisher` adapters_"?

Reply:
Relevant code (since refactoring) [is here](https://github.com/micronaut-projects/micronaut-core/blob/332cfd3278089dd5d14a245fe91609f0cae82f6e/http-client-jdk/src/main/java/io/micronaut/http/client/jdk/JdkHttpClient.java#L131-L152)